File: nova-api/src/main/java/xyz/xenondevs/nova/api/tileentity/TileEntity.java
package xyz.xenondevs.nova.api.tileentity;

import org.bukkit.Location;
import org.bukkit.OfflinePlayer;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import xyz.xenondevs.nova.api.block.NovaBlock;
import xyz.xenondevs.nova.api.material.NovaMaterial;

import java.util.List;

public interface TileEntity {
    
    /**
     * Gets the owner of this {@link TileEntity}.
     *
     * @return The owner of this {@link TileEntity}
     */
    @Nullable OfflinePlayer getOwner();
    
    /**
     * Gets the material of this {@link TileEntity}.
     *
     * @return The material of this {@link TileEntity}
     * @deprecated Use {@link #getBlock()} instead.
     */
    @Deprecated
    @NotNull NovaMaterial getMaterial();
    
    /**
     * Gets the {@link NovaBlock} of this {@link TileEntity}.
     *
     * @return The {@link NovaBlock} of this {@link TileEntity}
     */
    @NotNull NovaBlock getBlock();
    
    /**
     * Gets the {@link Location} of this {@link TileEntity}.
     *
     * @return The {@link Location} of this {@link TileEntity}
     */
    @NotNull Location getLocation();
    
    /**
     * Retrieves a list of all {@link ItemStack ItemStacks} this {@link TileEntity} would drop.
     *
     * @return A list of all {@link ItemStack ItemStacks} this {@link TileEntity} would drop.
     */
    @NotNull List<@NotNull ItemStack> getDrops(boolean includeSelf);
    
}

File: nova-api/src/main/kotlin/xyz/xenondevs/nova/api/tileentity/TileEntity.kt
package xyz.xenondevs.nova.api.tileentity

import org.bukkit.Location
import org.bukkit.OfflinePlayer
import org.bukkit.inventory.ItemStack
import xyz.xenondevs.nova.api.block.NovaBlock
import xyz.xenondevs.nova.api.material.NovaMaterial

@Suppress("DEPRECATION")
interface TileEntity {
    
    /**
     * The owner of this [TileEntity]
     */
    val owner: OfflinePlayer?
    
    /**
     * The material of this [TileEntity]
     */
    @Deprecated("Use NovaBlock instead", ReplaceWith("block"))
    val material: NovaMaterial
    
    /**
     * The block type of this [TileEntity]
     */
    val block: NovaBlock
    
    /**
     * The location of this [TileEntity]
     */
    val location: Location
    
    /**
     * Retrieves a list of all [ItemStacks][ItemStack] this [TileEntity] would drop
     */
    fun getDrops(includeSelf: Boolean): MutableList<ItemStack>
    
}